# Case Study

## Part 1
*To be completed at the conclusion of Day 1*

For the following exercises, you should use the data stored at `../data/companies.csv`
You aren't expected to finish all the exercises; just get through as many as time allows and we will review them together.

1. Start by becoming familiar with the data. How many rows and how many columns does it have? What are the data types of the columns?

In [1]:
import pandas as pd
comp_df = pd.read_csv('../data/companies.csv')
comp_df.head()

Symbol                 Name                  Sector
0    MMM           3M Company             Industrials
1    AOS      A.O. Smith Corp             Industrials
2    ABT  Abbott Laboratories             Health Care
3   ABBV          AbbVie Inc.             Health Care
4    ACN        Accenture plc  Information Technology

In [2]:
comp_df.shape

(505, 3)

In [3]:
comp_df.columns

Index(['Symbol', 'Name', 'Sector'], dtype='object')

In [4]:
comp_df.dtypes

Symbol    object
Name      object
Sector    object
dtype: object

2. Set the data's index to be the "Symbol" column.

In [5]:
#comp_df index doesn't change
#we're creating a new data frame that has Symbol as the index
#new data frame is allowed to have the same name as the original, if you want
comp_sym_df = comp_df.set_index('Symbol')
comp_sym_df.head()

Name                  Sector
Symbol                                             
MMM              3M Company             Industrials
AOS         A.O. Smith Corp             Industrials
ABT     Abbott Laboratories             Health Care
ABBV            AbbVie Inc.             Health Care
ACN           Accenture plc  Information Technology

3. Look up the company with the symbol NCLH. What company is this? What sector is it in?

In [6]:
#filter the original data frame using a condition
#returns a boolean value for each row
comp_df_filter = comp_df['Symbol'] == 'NCLH'
comp_df[comp_df_filter].head()

Symbol                   Name                  Sector
343   NCLH  Norwegian Cruise Line  Consumer Discretionary

In [20]:
#or slice using the Symbol index in the new data frame
#returns object... just the data from the requested slice
comp_sym_df.loc['NCLH']

Name       Norwegian Cruise Line
Sector    Consumer Discretionary
Name: NCLH, dtype: object

In [7]:
#or slice using the Symbol index in the new data frame
#returns data frame... just the data from the requested slice
comp_sym_df.loc[['NCLH']]

Name                  Sector
Symbol                                               
NCLH    Norwegian Cruise Line  Consumer Discretionary

4. Filter down to companies that *either* in the "Consumer Discretionary" or the "Consumer Staples" sectors.

In [8]:
sect_1 = comp_df['Sector'] == 'Consumer Discretionary'
sect_2 = comp_df['Sector'] == 'Consumer Staples'
sect_1or2 = sect_1 | sect_2
sect_1or2

0      False
1      False
2      False
3      False
4      False
       ...  
500    False
501     True
502    False
503    False
504    False
Name: Sector, Length: 505, dtype: bool

5. How many companies are left in the data now?

In [9]:
sum(sect_1or2)

116

In [10]:
comp_df[sect_1or2]

Symbol                       Name                  Sector
8      AAP         Advance Auto Parts  Consumer Discretionary
29      MO           Altria Group Inc        Consumer Staples
30    AMZN            Amazon.com Inc.  Consumer Discretionary
53    APTV                  Aptiv Plc  Consumer Discretionary
54     ADM  Archer-Daniels-Midland Co        Consumer Staples
..     ...                        ...                     ...
481    WBA   Walgreens Boots Alliance        Consumer Staples
491    WHR            Whirlpool Corp.  Consumer Discretionary
494    WYN          Wyndham Worldwide  Consumer Discretionary
495   WYNN           Wynn Resorts Ltd  Consumer Discretionary
501    YUM            Yum! Brands Inc  Consumer Discretionary

[116 rows x 3 columns]

6. Create a new column, "Symbol_Length", that is the length of the symbol of each company. *Hint: you may need to reset an index along the way.*

In [11]:
comp_df = comp_df.reset_index()
comp_df.head()

index Symbol                 Name                  Sector
0      0    MMM           3M Company             Industrials
1      1    AOS      A.O. Smith Corp             Industrials
2      2    ABT  Abbott Laboratories             Health Care
3      3   ABBV          AbbVie Inc.             Health Care
4      4    ACN        Accenture plc  Information Technology

In [12]:
symbol_len = comp_df['Symbol'].str.len()
symbol_len

0      3
1      3
2      3
3      4
4      3
      ..
500    3
501    3
502    3
503    4
504    3
Name: Symbol, Length: 505, dtype: int64

In [13]:
comp_df['Symbol_Length'] = symbol_len
comp_df.head()

index Symbol                 Name                  Sector  Symbol_Length
0      0    MMM           3M Company             Industrials              3
1      1    AOS      A.O. Smith Corp             Industrials              3
2      2    ABT  Abbott Laboratories             Health Care              3
3      3   ABBV          AbbVie Inc.             Health Care              4
4      4    ACN        Accenture plc  Information Technology              3

7. Find the company named "Kroger Co.". Change its name to "The Kroger Company".

In [14]:
comp_df.loc[comp_df['Name'] == 'Kroger Co.'].head()

index Symbol        Name            Sector  Symbol_Length
275    275     KR  Kroger Co.  Consumer Staples              2

In [15]:
comp_df.loc[(comp_df.Name == 'Kroger Co.'),'Name']='The Kroger Company'
comp_df.loc[comp_df['Name'] == 'The Kroger Company'].head()

index Symbol                Name            Sector  Symbol_Length
275    275     KR  The Kroger Company  Consumer Staples              2

In [22]:
#above works without the .loc
#per instructor, it's better to use .loc syntax
#so python knows to pull back the whole row of data
comp_df[comp_df['Name'] == 'The Kroger Company'].head()

index Symbol                Name            Sector  Symbol_Length
275    275     KR  The Kroger Company  Consumer Staples              2

In [16]:
#lines above work fine
#trying another approach to see if it works
import pandas as pd
comp_df_2 = pd.read_csv('../data/companies.csv')
comp_df_2.head()

comp_df_kroger = comp_df_2['Name'] == 'Kroger Co.'
comp_df_2[comp_df_kroger].head()

Symbol        Name            Sector
275     KR  Kroger Co.  Consumer Staples

In [17]:
#numpy approach works too!
import numpy as np
comp_df_2['Name'] = np.where((comp_df_2.Name == 'Kroger Co.'),'The Kroger Company',comp_df_2.Name)
comp_df_2[comp_df_kroger].head()

Symbol                Name            Sector
275     KR  The Kroger Company  Consumer Staples

**Bonus**: *For these two exercises, you won't find examples of the solution in our notebooks.
You'll need to search for help on the internet.*

*Don't worry if you aren't able to solve them.*

1. Filter down to companies whose symbol starts with A. How many companies meet this criterion?
2. What is the longest company name remaining in the dataset? You could just search the data visually, but try to find a programmatic solution.

In [18]:
filtered = comp_df['Name'].str.match(pat = '^A')
filtered

0      False
1       True
2       True
3       True
4       True
       ...  
500    False
501    False
502    False
503    False
504    False
Name: Name, Length: 505, dtype: bool

In [19]:
comp_df[filtered].head()

index Symbol                 Name                  Sector  Symbol_Length
1      1    AOS      A.O. Smith Corp             Industrials              3
2      2    ABT  Abbott Laboratories             Health Care              3
3      3   ABBV          AbbVie Inc.             Health Care              4
4      4    ACN        Accenture plc  Information Technology              3
5      5   ATVI  Activision Blizzard  Information Technology              4

In [23]:
#instructors showed this approach
filtered2 = comp_df['Name'].str.startswith(pat = 'A')
filtered2

0      False
1       True
2       True
3       True
4       True
       ...  
500    False
501    False
502    False
503    False
504    False
Name: Name, Length: 505, dtype: bool

In [24]:
comp_df[filtered2].head()

index Symbol                 Name                  Sector  Symbol_Length
1      1    AOS      A.O. Smith Corp             Industrials              3
2      2    ABT  Abbott Laboratories             Health Care              3
3      3   ABBV          AbbVie Inc.             Health Care              4
4      4    ACN        Accenture plc  Information Technology              3
5      5   ATVI  Activision Blizzard  Information Technology              4

## Part 2
*To be completed at the conclusion of Day 2*

This section again uses the data at `../data/companies.csv`.

1. Re-create the "Symbol_Length" column (see above).

In [2]:
import pandas as pd
comp_df = pd.read_csv('../data/companies.csv')
comp_df

Symbol                    Name                  Sector
0      MMM              3M Company             Industrials
1      AOS         A.O. Smith Corp             Industrials
2      ABT     Abbott Laboratories             Health Care
3     ABBV             AbbVie Inc.             Health Care
4      ACN           Accenture plc  Information Technology
..     ...                     ...                     ...
500    XYL              Xylem Inc.             Industrials
501    YUM         Yum! Brands Inc  Consumer Discretionary
502    ZBH  Zimmer Biomet Holdings             Health Care
503   ZION           Zions Bancorp              Financials
504    ZTS                  Zoetis             Health Care

[505 rows x 3 columns]

2. What is the average symbol length of companies in the data set?

In [8]:
symbol_len = comp_df['Symbol'].str.len()
symbol_len

0      3
1      3
2      3
3      4
4      3
      ..
500    3
501    3
502    3
503    4
504    3
Name: Symbol, Length: 505, dtype: int64

In [10]:
comp_df['symbol_len'] = symbol_len
comp_df

Symbol                    Name                  Sector  symbol_len
0      MMM              3M Company             Industrials           3
1      AOS         A.O. Smith Corp             Industrials           3
2      ABT     Abbott Laboratories             Health Care           3
3     ABBV             AbbVie Inc.             Health Care           4
4      ACN           Accenture plc  Information Technology           3
..     ...                     ...                     ...         ...
500    XYL              Xylem Inc.             Industrials           3
501    YUM         Yum! Brands Inc  Consumer Discretionary           3
502    ZBH  Zimmer Biomet Holdings             Health Care           3
503   ZION           Zions Bancorp              Financials           4
504    ZTS                  Zoetis             Health Care           3

[505 rows x 4 columns]

In [11]:
comp_df.agg({
    'symbol_len': ['mean']
})

symbol_len
mean    3.122772

3. What is the average symbol length by sector? That is, after grouping by sector, what is the average symbol length for each group?

In [12]:
comp_df.groupby('Sector').agg({'symbol_len': ['mean']}).head()

symbol_len
                             mean
Sector                           
Consumer Discretionary   3.170732
Consumer Staples         2.882353
Energy                   3.032258
Financials               3.086957
Health Care              3.278689

4. How long is the longest company name? How long is the longest company name by sector?

In [13]:
comp_name_len = comp_df['Name'].str.len()
comp_name_len

0      10
1      15
2      19
3      11
4      13
       ..
500    10
501    15
502    22
503    13
504     6
Name: Name, Length: 505, dtype: int64

In [15]:
comp_df['comp_name_len'] = comp_name_len
comp_df

Symbol                    Name                  Sector  symbol_len  \
0      MMM              3M Company             Industrials           3   
1      AOS         A.O. Smith Corp             Industrials           3   
2      ABT     Abbott Laboratories             Health Care           3   
3     ABBV             AbbVie Inc.             Health Care           4   
4      ACN           Accenture plc  Information Technology           3   
..     ...                     ...                     ...         ...   
500    XYL              Xylem Inc.             Industrials           3   
501    YUM         Yum! Brands Inc  Consumer Discretionary           3   
502    ZBH  Zimmer Biomet Holdings             Health Care           3   
503   ZION           Zions Bancorp              Financials           4   
504    ZTS                  Zoetis             Health Care           3   

     comp_nam_len  comp_name_len  
0              10             10  
1              15             15  
2              19             19  
3              11             11  
4              13             13  
..            ...            ...  
500            10             10  
501            15             15  
502            22             22  
503            13             13  
504             6              6  

[505 rows x 6 columns]

In [20]:
comp_df_name = comp_df[['Name', 'comp_name_len']]
comp_df_name

Name  comp_name_len
0                3M Company             10
1           A.O. Smith Corp             15
2       Abbott Laboratories             19
3               AbbVie Inc.             11
4             Accenture plc             13
..                      ...            ...
500              Xylem Inc.             10
501         Yum! Brands Inc             15
502  Zimmer Biomet Holdings             22
503           Zions Bancorp             13
504                  Zoetis              6

[505 rows x 2 columns]

In [22]:
comp_df_name.sort_values(by = ('comp_name_len'), ascending=False)

Name  comp_name_len
191  Fidelity National Information Services             38
278     Laboratory Corp. of America Holding             35
19      Alexandria Real Estate Equities Inc             35
35       American International Group, Inc.             34
436       The Bank of New York Mellon Corp.             33
..                                      ...            ...
245                                  Incyte              6
350                                   ONEOK              5
46                                    ANSYS              5
382                                   Qorvo              5
336                                    Nike              4

[505 rows x 2 columns]

Now open the pricing data at `../data/prices.csv`.
Note that this data is entirely fabricated and does not exhibit the qualities of real stock market data!

1. Become familiar with this data. What is its shape? What are its data types?

In [32]:
import pandas as pd
prices_df = pd.read_csv('../data/prices.csv')
prices_df.head()

Symbol       Price  Quarter
0    MMM  150.259437        1
1    AOS  103.668371        1
2    ABT  253.780326        1
3   ABBV  306.823770        1
4    ACN  211.437491        1

In [ ]:
prices_df.shape

In [27]:
prices_df.dtypes

Symbol      object
Price      float64
Quarter      int64
dtype: object

2. Get summary metrics (count, min, max, standard deviation, etc) for both the Price and Quarter columns. *Hint: we saw a method of DataFrames that will do this for you in a single line.*

In [24]:
prices_df.describe()

Price      Quarter
count  2020.000000  2020.000000
mean    219.446042     2.500000
std      85.912103     1.118311
min      45.437073     1.000000
25%     149.248976     1.750000
50%     215.679389     2.500000
75%     286.468536     3.250000
max     394.017364     4.000000

3. Perform an inner join between this data set and the companies data, on the Symbol column.

In [28]:
prices_df.columns.intersection(comp_df.columns)

Index(['Symbol'], dtype='object')

In [30]:
prices_comp_df = pd.merge(prices_df, comp_df, left_on = 'Symbol', right_on = 'Symbol')
prices_comp_df

Symbol       Price  Quarter             Name       Sector  symbol_len  \
0       MMM  150.259437        1       3M Company  Industrials           3   
1       MMM  134.425207        2       3M Company  Industrials           3   
2       MMM  135.013910        3       3M Company  Industrials           3   
3       MMM  137.325815        4       3M Company  Industrials           3   
4       AOS  103.668371        1  A.O. Smith Corp  Industrials           3   
...     ...         ...      ...              ...          ...         ...   
2015   ZION   95.158470        4    Zions Bancorp   Financials           4   
2016    ZTS  369.255493        1           Zoetis  Health Care           3   
2017    ZTS  362.047214        2           Zoetis  Health Care           3   
2018    ZTS  369.518002        3           Zoetis  Health Care           3   
2019    ZTS  359.106033        4           Zoetis  Health Care           3   

      comp_nam_len  comp_name_len  
0               10             10  
1               10             10  
2               10             10  
3               10             10  
4               15             15  
...            ...            ...  
2015            13             13  
2016             6              6  
2017             6              6  
2018             6              6  
2019             6              6  

[2020 rows x 8 columns]

4. How many rows does our data have now?

In [31]:
prices_comp_df.shape

(2020, 8)

5. What do you think this data represents? Form a hypothesis and look through the data more carefully until you are confident you understand what it is and how it is structured.

6. Group the data by sector. What is the average first quarter price for a company in the Real Estate sector? What is the minimum fourth quarter price for a company in the Industrials sector?

In [75]:
pc_sect_quart_df = prices_comp_df.groupby(['Sector', 'Quarter'], as_index = False).agg({'Price': ['mean', 'min']})
filter_1 = pc_sect_quart_df['Sector'] == 'Real Estate'
filter_2 = pc_sect_quart_df['Sector'] == 'Industrials'
filter_3 = pc_sect_quart_df['Quarter'] == 1
filter_4 = pc_sect_quart_df['Quarter'] == 4
pc_sect_quart_filtered_df = (filter_1 & filter_3) | (filter_2 & filter_4)
pc_sect_quart_df[pc_sect_quart_filtered_df]

Sector Quarter       Price           
                               mean        min
23  Industrials       4  205.150942  56.452515
32  Real Estate       1  232.728266  78.208964

7. Filter the data down to just prices for Apple, Google, Microsoft, and Amazon.

In [70]:
comp_df.head()

Symbol                 Name                  Sector  symbol_len  \
0    MMM           3M Company             Industrials           3   
1    AOS      A.O. Smith Corp             Industrials           3   
2    ABT  Abbott Laboratories             Health Care           3   
3   ABBV          AbbVie Inc.             Health Care           4   
4    ACN        Accenture plc  Information Technology           3   

   comp_nam_len  comp_name_len  
0            10             10  
1            15             15  
2            19             19  
3            11             11  
4            13             13

In [57]:
comp_df[comp_df['Symbol'].str.match(pat = '^AMZN')]

Symbol             Name                  Sector  symbol_len  comp_nam_len  \
30   AMZN  Amazon.com Inc.  Consumer Discretionary           4            15   

    comp_name_len  
30             15

In [58]:
comp_df[comp_df['Symbol'].str.match(pat = '^GO')]

Symbol                  Name                  Sector  symbol_len  \
27  GOOGL  Alphabet Inc Class A  Information Technology           5   
28   GOOG  Alphabet Inc Class C  Information Technology           4   

    comp_nam_len  comp_name_len  
27            20             20  
28            20             20

In [64]:
comp_df[comp_df['Name'].str.match(pat = '^Mic')]

Symbol                   Name                  Sector  symbol_len  \
309   KORS  Michael Kors Holdings  Consumer Discretionary           4   
310   MCHP   Microchip Technology  Information Technology           4   
311     MU      Micron Technology  Information Technology           2   
312   MSFT        Microsoft Corp.  Information Technology           4   

     comp_nam_len  comp_name_len  
309            21             21  
310            20             20  
311            17             17  
312            15             15

In [65]:
comp_df[comp_df['Name'].str.match(pat = '^App')]

Symbol                    Name                  Sector  symbol_len  \
51   AAPL              Apple Inc.  Information Technology           4   
52   AMAT  Applied Materials Inc.  Information Technology           4   

    comp_nam_len  comp_name_len  
51            10             10  
52            22             22

In [69]:
#grabbed the symbols I found in lines above this
prices_comp_df_filter = prices_comp_df['Symbol'].isin(['AAPL', 'GOOGL', 'GOOG', 'MSFT', 'AMZN'])
prices_comp_df[prices_comp_df_filter].sort_values(by = ('Symbol'), ascending=True)

Symbol       Price  Quarter                  Name  \
207    AAPL  266.069127        4            Apple Inc.   
206    AAPL  263.508526        3            Apple Inc.   
205    AAPL  269.961337        2            Apple Inc.   
204    AAPL  275.197947        1            Apple Inc.   
121    AMZN  229.964972        2       Amazon.com Inc.   
120    AMZN  220.750078        1       Amazon.com Inc.   
122    AMZN  233.972278        3       Amazon.com Inc.   
123    AMZN  241.194174        4       Amazon.com Inc.   
112    GOOG  216.432380        1  Alphabet Inc Class C   
113    GOOG  228.992923        2  Alphabet Inc Class C   
114    GOOG  232.580066        3  Alphabet Inc Class C   
115    GOOG  244.167211        4  Alphabet Inc Class C   
108   GOOGL  109.175657        1  Alphabet Inc Class A   
111   GOOGL   96.692314        4  Alphabet Inc Class A   
110   GOOGL   80.795190        3  Alphabet Inc Class A   
109   GOOGL   93.864951        2  Alphabet Inc Class A   
1248   MSFT  202.640689        1       Microsoft Corp.   
1249   MSFT  202.093615        2       Microsoft Corp.   
1250   MSFT  212.749636        3       Microsoft Corp.   
1251   MSFT  220.409889        4       Microsoft Corp.   

                      Sector  symbol_len  comp_nam_len  comp_name_len  
207   Information Technology           4            10             10  
206   Information Technology           4            10             10  
205   Information Technology           4            10             10  
204   Information Technology           4            10             10  
121   Consumer Discretionary           4            15             15  
120   Consumer Discretionary           4            15             15  
122   Consumer Discretionary           4            15             15  
123   Consumer Discretionary           4            15             15  
112   Information Technology           4            20             20  
113   Information Technology           4            20             20  
114   Information Technology           4            20             20  
115   Information Technology           4            20             20  
108   Information Technology           5            20             20  
111   Information Technology           5            20             20  
110   Information Technology           5            20             20  
109   Information Technology           5            20             20  
1248  Information Technology           4            15             15  
1249  Information Technology           4            15             15  
1250  Information Technology           4            15             15  
1251  Information Technology           4            15             15

8. Save this data as big_4.csv in the `../data` directory.

9. Using Seaborn, plot the price of these companies over 4 quarters. Encode the quarter as the x-axis, the price as the y-axis, and the company symbol as the hue.

**Bonus**:

This data is in a form that is useful for plotting.
But in this shape, it would be quite difficult to calculate the difference between each company's fourth quarter price and its first quarter price.

Reshape this data so it is of a form like the below:

| Symbol | Name | Sector | Q1 | Q2 | Q3 | Q4 |
|--------|------|--------|----|----|----|----|
| AAPL   | Apple Inc. | Information Technology | 275.20 | 269.96 | 263.51 | 266.07

From which we could easily calculate Q4 - Q1.

*You will probably want to google something like "python reshaping data". This is a very challenging problem!*

per Ethan, use the pivot method to do this